http/server.ts

```diff
@@ -264,114 +248,83 @@ export class ServerRequest {
   }
 }
 
-interface ServeEnv {
-  reqQueue: ServerRequest[];
-  serveDeferred: Deferred;
-}
-
 /** Continuously read more requests from conn until EOF
- * Calls maybeHandleReq.
  * bufr is empty on a fresh TCP connection.
  * Would be passed around and reused for later request on same conn
  * TODO: make them async function after this change is done
  * https://github.com/tc39/ecma262/pull/1250
  * See https://v8.dev/blog/fast-async
  */
-async function readRequest(
-  c: HttpConn,
-  bufr?: BufReader
-): Promise<[ServerRequest, BufState]> {
-  if (!bufr) {
-    bufr = new BufReader(c);
-  }
+async function* iterateHttpRequests(
+  c: HttpConn
+): AsyncIterableIterator<[ServerRequest | null, BufState]> {
+  const bufr = new BufReader(c);
   const bufw = new BufWriter(c);
-  const req = new ServerRequest();
-
-  // Set and incr pipeline id;
-  req.pipelineId = ++c.lastPipelineId;
-  // Set a new pipeline deferred associated with this request
-  // for future requests to wait for.
-  c.pendingDeferredMap.set(req.pipelineId, deferred());
-
-  req.conn = c;
-  req.r = bufr!;
-  req.w = bufw;
-  const tp = new TextProtoReader(bufr!);
-
-  let s: string;
-  let err: BufState;
-
-  // First line: GET /index.html HTTP/1.0
-  [s, err] = await tp.readLine();
-  if (err) {
-    return [null, err];
-  }
-  [req.method, req.url, req.proto] = s.split(" ", 3);
-
-  [req.headers, err] = await tp.readMIMEHeader();
 
-  return [req, err];
-}
+  for (;;) {
+    const req = new ServerRequest();
+
+    // Set and incr pipeline id;
+    req.pipelineId = ++c.lastPipelineId;
+    // Set a new pipeline deferred associated with this request
+    // for future requests to wait for.
+    c.pendingDeferredMap.set(req.pipelineId, deferred());
+
+    req.conn = c;
+    req.r = bufr!;
+    req.w = bufw;
+
+    // First line: GET /index.html HTTP/1.0
+    const tp = new TextProtoReader(bufr!);
```

Comment:
TextProtoReader construction could probably be moved to the top of the generator.

Reply:
No, it should be constructed per header. I've moved it back.